### PR TITLE
Update AutorecoveringConnection to call ShutdownListeners before reco…

### PR DESCRIPTION
**Pull request by doswell**
…nnecting.

Re; #135

### CODE REVIEW

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java [hunk 1]
```diff
@@ -456,9 +455,8 @@ synchronized private void beginAutomaticRecovery() throws InterruptedException,
     }
 
     private void recoverShutdownListeners() {
-        for (ShutdownListener sh : this.shutdownHooks) {
-            this.delegate.addShutdownListener(sh);
-        }
+    	//Add the auto recovering listener to the new delegate.
+    	addAutomaticRecoveryListener();
```
Comment (michaelklishin) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: I find it non-obvious how the (user-provided) hooks are recovered.

**Comment by michaelklishin** [comment]
I need to think if I'm 100% happy with this particular implementation of the idea in #135.

**Comment by doswell** [comment]
@michaelklishin  I'd probably say "no not happy" I've worked on a more clear, and probably more importantly, more consistent implementation of calling the user hooks. 
In short there is a ForwardingRecoveryListener" class where user hooks are added into, and which gets fired prior to the recovery starting. It also applies the same sync blocks that the existing delegate listeners use. I just have to get it up to git. 
The hooks are also just maintained within AutorecoveringConnection and not added to the underlying delegate.
I'll try and get that up tomorrow

**Comment by michaelklishin** [comment]
@doswell thank you. One idea I had is simply subclassing `ShutdownListener` (to produce a marker class) so that we can tell a recovery shutdown hooks from others so that they can be filtered out. How does that sound?

Alternatively we could use annotations but then I'm afraid this cannot be backported into `stable`.

**Comment by doswell** [comment]
I didn't see a need for the subclassing in terms of the shutdown listeners just notifying that a disconnect occurs. ie, keeping separate from #132. But I think the idea of subclassing may be useful for #132 in terms of having a RecoveryStartedListener / RecoveryCompletedListener which can register on the RecoveryListeners without needing API changes. 

Close for #141

**Comment by michaelklishin** [comment]
Thank you for your time. There are some class responsibility and cosmetic things in this PRs that I disagree with but I think I understand the intent in #135 well enough now. I will look into an alternative PR for #135.

**Pull request closed without merging**